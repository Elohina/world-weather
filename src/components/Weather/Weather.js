import React, { useState, useEffect, Fragment } from 'react';
import PropTypes from 'prop-types';
import _ from 'lodash';
import CardDetail from '../CardDetail/CardDetail';
import { fetchForecast } from '../../api/weatherApi';
import styled from 'styled-components';

import Forecast from '../Forecast/Forecast';

const Container = styled.div`
  display: flex;
  width: 100%;
  justify-content: center;

  @media(max-width: 75em) {
    flex-direction: column;
  }
`;

const Weather = ({ lat, lng, address }) => {
  const [forecast, setForecast] = useState({});
  const [selectedCard, setSelectedCard] = useState({});
  const [selectedDay, setSelectedDay] = useState({});
  const [showErrorMessage, setShowErrorMessage] = useState(false);

  useEffect(() => {
    const fetchData = async () => {
      try {
        const result = await fetchForecast(lat, lng);
        const mapped_days = _.mapValues(result, (element) => {
          return {
            ...element,
            date_label: new Date(element.dt * 1000).getUTCDate(),
          };
        });

        const weather_days = _.mapValues(_.groupBy(mapped_days, "date_label"));
        setForecast(weather_days);
        setSelectedCard(mapped_days[0]);
        setSelectedDay(weather_days[mapped_days[0].date_label]);
        setShowErrorMessage(false);
      } catch (error) {
        setShowErrorMessage(true);
      }

    };

    fetchData();
  }, [lat, lng]);

  const isToday = (date) => {
    const today = new Date().getUTCDate();
    return today === date;
  };

  const handleOnSelectCard = (card, day) => {
    setSelectedCard(card);
    setSelectedDay(forecast[day]);
  }

  return (
    <Container>
      { showErrorMessage ? (<p>Oops! There was a problem fetching the weather. Please try again later.</p>)
      : (
        <Fragment>
          <CardDetail
            selectedDay={selectedDay}
            selected_card={selectedCard}
            isToday={isToday(selectedCard.date_label)}
            address={address}
          />
          <Forecast
            forecast={forecast}
            onChangeTime={(card, day)=>{handleOnSelectCard(card, day);}}
            isToday={isToday}
            selectedDay={selectedDay}
          />
        </Fragment>
      )}
    </Container>
  );
};

export default Weather;

Weather.propTypes = {
  lan: PropTypes.string,
  lng: PropTypes.string,
  address: PropTypes.string
}
