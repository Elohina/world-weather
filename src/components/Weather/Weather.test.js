import React from "react";
import { mount, shallow } from "enzyme";
import Weather from "./Weather.js";

describe("<Weather />", () => {
  const wrapper = mount(<Weather lat={"53.3498053"} lng={"-6.2603097"} address={"Dublin, Ireland"} />);
  
  it('should match the snapshot', () => {
    expect(wrapper.html()).toMatchSnapshot();
  });

  it("should render component correctly", () => {
    expect(wrapper.exists()).toBe(true);
    expect(wrapper.find('CardDetail').length).toBe(1);
    expect(wrapper.find('Forecast').length).toBe(1);
  });
});