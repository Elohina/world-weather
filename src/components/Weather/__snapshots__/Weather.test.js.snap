// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Weather /> should match the snapshot 1`] = `"<div class=\\"sc-fzqBZW dWMMkj\\"><div class=\\"sc-AxhCb DnkkQ\\"></div><div class=\\"sc-fzokOt bOMXGc\\"><h3>Week</h3><h5>Click on each day to see the details</h5></div></div>"`;
